src/vivarium_public_health/metrics/stratification.py

```diff
@@ -0,0 +1,353 @@
+    current_categories_getter: Callable[[], Set[str]] = None
+
+    def __post_init__(self):
+        self.mapper = self.mapper if self.mapper else self._default_mapper
+        self.current_categories_getter = (
+            self.current_categories_getter if self.current_categories_getter
+            else self._default_current_categories_getter
+        )
+
+    def get_current_categories(self) -> Set[str]:
+        return self.current_categories_getter()
+
+    def _default_mapper(self, row: pd.Series) -> str:
+        category = str(row[0])
+        if category not in self.categories:
+            raise ValueError(f"Invalid value '{category}' found in {self.name}.")
+        return category
+
+    def _default_current_categories_getter(self) -> Set[str]:
+        return self.categories
+
+
+class ResultsStratifier:
+    """Centralized component for handling results stratification.
+
+    This should be used as a subcomponent for observers.  The observers
+    can then ask this component for population subgroups and labels during
+    results production and have this component manage adjustments to the
+    final column labels for the subgroups.
+
+    """
+
+    NAME = "results_stratifier"
+
+    configuration_defaults = {
+        "default": [],
+    }
+
+    def __init__(self):
+        self.configuration_defaults = self._get_configuration_defaults()
+
+        self.metrics_pipeline_name = "metrics"
+        self.tmrle_key = "population.theoretical_minimum_risk_life_expectancy"
+
+    ##########################
+    # Initialization methods #
+    ##########################
+
+    # noinspection PyMethodMayBeStatic
+    def _get_configuration_defaults(self) -> Dict[str, Dict]:
+        return {
+            "observers": {
+                "default": ResultsStratifier.configuration_defaults["default"]
+            }
+        }
+
+    ##############
+    # Properties #
+    ##############
+
+    @property
+    def name(self) -> str:
+        return ResultsStratifier.NAME
+
+    #################
+    # Setup methods #
+    #################
+
+    # noinspection PyAttributeOutsideInit
+    def setup(self, builder: Builder) -> None:
+        """Perform this component's setup."""
+        self.clock = None
+        self.default_stratification_levels = self._get_default_stratification_levels(builder)
+        self.pipelines = {}
+        self.columns_required = ["tracked"]
+        self.clock_sources = set()
+        self.stratification_levels: Dict[str, StratificationLevel] = {}
+        self.stratification_groups: pd.DataFrame = None
+
+        self.age_bins = self._get_age_bins(builder)
+
+        self.register_stratifications(builder)
+        self.population_view = self._get_population_view(builder)
+
+        self._register_timestep_prepare_listener(builder)
+        self._register_simulation_end_listener(builder)
+
+    # noinspection PyMethodMayBeStatic
+    def _get_clock(self, builder: Builder) -> Callable[[], Time]:
```

Comment:
Just for thought on the structure of components and their inheritibility:

The layer of indirection that you make to access builder services is a signal about what is available and potentially _intended_ to be modified during inheritance.  The thing to keep in mind then, is that we have several kinds of things (e.g. the timekeeping system) that are meant to be modified by configuration and not by inheritance.

Reply:
Good point. I'll consider what things can be changed in light of that observation.